hazelcast/src/main/java/com/hazelcast/cache/impl/TenantContextual.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.cache.impl;
+
+import com.hazelcast.spi.tenantcontrol.TenantControl;
+import com.hazelcast.spi.tenantcontrol.TenantControl.Closeable;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.function.Supplier;
+
+/**
+ * Represents a value that requires tenant control context to be accessed
+ *
+ * @author lprimak
+ * @param <T> object type
+ */
+public class TenantContextual<T> {
+    private T contextual;
+    private volatile boolean initialized;
+    private final Supplier<T> initFunction;
+    private final Supplier<Boolean> existsFunction;
+    private final TenantControl tenantControl;
+    private final Lock lock = new ReentrantLock();
+
+    public TenantContextual(Supplier<T> initFunction, Supplier<Boolean> existsFunction, TenantControl tenantControl) {
+        this.initFunction = initFunction;
+        this.existsFunction = existsFunction;
+        this.tenantControl = tenantControl;
+    }
+
+    /**
+     *
+     * @return underlying object, initialize within Tenant Control when necessary
+     */
+    public T get() {
+        boolean localInitialized = this.initialized;
+        if (!localInitialized) {
+            try {
+                lock.lock();
```

Comment:
Calling allien code while holding a lock is a source or problems.

Reply:
I don't believe any alien code is called from here. This class is really private API (perhaps I should annotate it as such) but the way it's used only "non-alien" code is used here

Reply:
Alien code -> Code that isn't owned by this class.  The problem with any interface injected into this class and called whole holding a lock, it is very easy to run into severe problems (contention, deadlocks etc). And perhaps it is fine now, but I have seen on quite a few occasions that eventually you end up with some nasty locking problem.

Reply:
This should be called once, before the tenant context is initialised. Probably isn't going to be a problem. @lprimak is it possible to initialise the tenant control eagerly while record store is being created and not later when it's accessed?

Reply:
no, as when Hazelcast migration happens, no user classes (factory classes here) are available so this will result in errors

Reply:
Resolving as it should be lazy and it will be invoked only once.